cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/kafkaassigner/KafkaAssignerDiskUsageDistributionGoal.java

```diff
@@ -0,0 +1,568 @@
+   */
+  Replica findReplicaToSwapWith(Replica replica,
+                                List<ReplicaWrapper> sortedReplicasToSearch,
+                                double targetSize,
+                                double minSize,
+                                double maxSize,
+                                ClusterModel clusterModel) {
+    int pos = findReplicaPos(sortedReplicasToSearch, targetSize, 0);
+    int minPos = findReplicaPos(sortedReplicasToSearch, minSize, 1);
+    int maxPos = findReplicaPos(sortedReplicasToSearch, maxSize, -1);
+    if (minPos > maxPos) {
+      return null;
+    }
+    // It is possible that the target size is out of the range of minSize and maxSize. In that case, we make it become
+    // the closest one in range.
+    pos = Math.max(pos, minPos);
+    pos = Math.min(pos, maxPos);
+
+    // The following logic starts from pos and searches higher and lower position until it finds a replica that 
+    // is eligible to swap with.
+    int low = pos;
+    int high = pos;
+    while (pos >= minPos && pos <= maxPos) {
+      Replica toSwapWith = sortedReplicasToSearch.get(pos).replica();
+      if (canSwap(replica, sortedReplicasToSearch.get(pos).replica(), clusterModel)) {
+        // found the candidate.
+        return toSwapWith;
+      } else {
+        // get the next position.
+        pos = findNextPos(sortedReplicasToSearch, targetSize, low - 1, high + 1, minPos, maxPos);
+        if (pos == low - 1) {
+          low--;
+        } else {
+          high++;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Find the next position between the provided low and high position, whichever is closer to the target size.
+   * 
+   * @param sortedReplicas the replica list sorted in ascending order based on size.
+   * @param targetSize the target size of replica to search
+   * @param low the low pointer candidate
+   * @param high the high pointer candidate
+   * @param minPos the minimum allowed position
+   * @param maxPos the maximum allowed position
+   *               
+   * @return the position chosen between high and low.
+   */
+  private int findNextPos(List<ReplicaWrapper> sortedReplicas, double targetSize, int low, int high, int minPos, int maxPos) {
+    if (low < minPos) {
+      return high;
+    } else if (high > maxPos) {
+      return low;
+    } else {
+      double leftSizeDiff = Math.abs(replicaSize(sortedReplicas.get(low).replica()) - targetSize);
+      double rightSizeDiff = Math.abs(replicaSize(sortedReplicas.get(high).replica()) - targetSize);
+      return leftSizeDiff <= rightSizeDiff ? low : high;
+    }
+  }
+
+  /**
+   * Checks whether the two replicas can swap with each other. Two replicas can swap if:
+   * 1. r1 and r2 are in the same rack (this assumes the initial assignment is already rack aware), OR
+   * 2. partition of r1 does not have replica in the rack of r2, and vice versa
+   * 
+   * In addition, r1 and r2 must have the same role.
```

Comment:
Should we clarify what `the same role` means -- i.e. either a leader or a follower in any position in partition replica list.